### PR TITLE
Optimize uiopacity component

### PR DESCRIPTION
**Pull request by qiuguohua**
Re: #

### Changelog

https://github.com/cocos/cocos-engine/issues/16514

document:
https://github.com/cocos/3d-tasks/discussions/18155

test-case:
https://github.com/cocos/cocos-test-projects/pull/857

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: cocos/2d/components/ui-opacity.ts [hunk 1]
```diff
@@ -106,50 +127,93 @@ export class UIOpacity extends Component {
 
         const render = node._uiProps.uiComp as UIRenderer;
         const uiOp = node.getComponent<UIOpacity>(UIOpacity);
-        let interruptOpacity = interruptParentOpacity;// if there is no UIOpacity component, it should always equal to 1.
 
+        if (recursiveWithoutUiOp && uiOp) {
```
Comment (minggo) [inline cocos/2d/components/ui-opacity.ts]: `recursiveWithoutUiOp ` means recursing only nodes without `uiopacity` component,  and `uiOp` means it has a `uiopacity` component. It seems conflict here.

File: cocos/2d/components/ui-opacity.ts [hunk 2]
```diff
@@ -96,8 +97,28 @@ export class UIOpacity extends Component {
...
+     * @zh
+     * 递归设置localopacity。
+     *
+     * @param node @en recursive node.
+     *             @zh 递归的节点。
+     * @param node @en Is the color dirty.
+     *             @zh color是否dirty。
+     * @param parentOpacity @en The parent node's opacity.
+     *                      @zh 父节点的opacity。
+     * @param recursiveWithoutUiOp @en Recursing only nodes without uiopacity.
```
Comment (minggo) [inline cocos/2d/components/ui-opacity.ts]: without uiopacity -> without `UIOpacity` component.

File: cocos/2d/components/ui-opacity.ts [hunk 3]
```diff
@@ -96,8 +97,28 @@ export class UIOpacity extends Component {
...
+     *                             @zh 仅仅递归没有uiopacity的组件。
```
Comment (minggo) [inline cocos/2d/components/ui-opacity.ts]: uiopacity -> UIOpacity

File: cocos/2d/components/ui-opacity.ts [hunk 4]
```diff
@@ -106,50 +127,93 @@ export class UIOpacity extends Component {
...
+            this._parentOpacity = this._getParentOpacity(parent);
+            opacity = this._parentOpacity;
+        }
+        UIOpacity.setEntityLocalOpacityDirtyRecursively(this.node, true, opacity, false);
+    }
+
+    protected _setEntityLocalOpacityRecursively (opacity: number): void {
+        if (!JSB) {
+            return;
+        }
```
Comment (minggo) [inline cocos/2d/components/ui-opacity.ts:195]: Why return if not JSB?

File: cocos/2d/components/ui-opacity.ts (hunk 1 above)
Comment (qiuguohua) [reply cocos/2d/components/ui-opacity.ts]: 'uiOp ! = null' is indicating that the UiOpacity component exists. ‘recursiveWithoutUiOp‘ indicates that it doesn't need to exist, so both conditions are met when handling the situation.

File: cocos/2d/components/ui-opacity.ts (hunk 2 above)
Comment (qiuguohua) [reply cocos/2d/components/ui-opacity.ts]: ok

File: cocos/2d/components/ui-opacity.ts (hunk 4 above)
Comment (qiuguohua) [reply cocos/2d/components/ui-opacity.ts:195]: This is only necessary for native

Comment (minggo) [reply cocos/2d/components/ui-opacity.ts:195]: Could you please add some comment why only need in native?

Comment (qiuguohua) [reply cocos/2d/components/ui-opacity.ts:195]: Because JSB's localOpacity value is present in the renderEntity, but non-JSB's are not.

**Comment by kaxifakl** [comment]
I think change 'walk' function in Batcher2D.cpp to the same logic in Batcher2D.ts is a good solution. 
Get opacity value from parent and just bind value with jsb on native.

**Comment by qiuguohua** [comment]
> I think change 'walk' function in Batcher2D.cpp to the same logic in Batcher2D.ts is a good solution.
Get opacity value from parent and just bind value with jsb on native.

Yes. This is a very good idea that can be moved on to the next stage.

**Review by minggo — APPROVED** [review]

**Pull request merged**